File: .config/ags/notifications.js
const notifications = await Service.import("notifications");

function NotificationIcon({ app_entry, app_icon, image }) {
  if (image) {
    return Widget.Box({
      css:
        `background-image: url("${image}");` +
        "background-size: contain;" +
        "background-repeat: no-repeat;" +
        "background-position: center;",
    });
  }

  let icon = "dialog-information-symbolic";
  if (Utils.lookUpIcon(app_icon)) icon = app_icon;

  if (app_entry && Utils.lookUpIcon(app_entry)) icon = app_entry;

  return Widget.Box({
    child: Widget.Icon(icon),
  });
}

function Notification(n) {
  const icon = Widget.Box({
    css: "padding: 10px",
    vpack: "start",
    class_name: "icon",
    child: NotificationIcon(n),
  });

  const title = Widget.Label({
    xalign: 0,
    justification: "left",
    hexpand: true,
    max_width_chars: 24,
    truncate: "end",
    wrap: true,
    label: n.summary,
    use_markup: true,
  });

  const body = Widget.Label({
    //css: "font-size: 40px",
    hexpand: true,
    use_markup: true,
    xalign: 0,
    justification: "left",
    label: n.body,
    wrap: true,
  });

  const actions = Widget.Box({
    class_name: "actions",
    children: n.actions.map(({ id, label }) =>
      Widget.Button({
        class_name: "action-button",
        on_clicked: () => {
          n.invoke(id);
          n.dismiss();
        },
        hexpand: true,
        child: Widget.Label(label),
      }),
    ),
  });

  return Widget.EventBox({
    attribute: { id: n.id },
    on_primary_click: n.dismiss,
    expand: true,
    vpack: "center",
    child: Widget.Box({
      css: "padding: 10px",
      hpack: "center",
      class_name: `notification ${n.urgency}`,
      vertical: true,
      children: [
        Widget.Box([icon, Widget.Box({ vertical: true }, title, body)]),
        actions,
      ],
    }),
  });
}

export default (monitor = 0) => {
  const list = Widget.Box({
    class_name: "container",
    vertical: true,
    expand: true,
    css: "min-width: 300px; min-height: 100px;" + "padding:10px",
    children: notifications.popups.map(Notification),
  });

  const rev = Widget.Revealer({
    child: list,
  });

  function onNotified(_, id) {
    const n = notifications.getNotification(id);
    if (n) {
      list.children = [Notification(n), ...list.children];
      rev.reveal_child = true;
    }
  }

  function onDismissed(_, id) {
    list.children.find((n) => n.attribute.id === id)?.destroy();
    rev.reveal_child = false;
  }

  list
    .hook(notifications, onNotified, "notified")
    .hook(notifications, onDismissed, "dismissed");

  return Widget.Window({
    layer: "overlay",
    monitor,
    name: `notifications${monitor}`,
    class_name: "notification",
    anchor: ["top"],
    child: Widget.Box({
      css: "padding: 10px;",
      vertical: true,
      child: rev,
    }),
  });
};

File: .config/astal/lua/widget/notifications.lua
local timeout = astal.timeout
local Notifd = astal.require("AstalNotifd")
local notifd = Notifd.get_default()
local popup_timeout = 3000
local Widget = require("astal.gtk3.widget")

local notif_item = function(n)
	local app_name = Widget.Label({
		label = n.app_name,
		class_name = "name",
	})

	local summary = Widget.Label({
		label = n.summary,
		class_name = "summary",
		halign = "START",
		valign = "CENTER",
		wrap = true,
		xalign = 0,
	})

	local body = Widget.Label({
		label = n.body,
		class_name = "body",
		halign = "START",
		valign = "START",
		wrap = true,
		xalign = 0,
		-- max_width_chars = 130,
	})

	local app_icon = Widget.Icon({
		icon = lookup_icon({ n.app_name, "application-x-executable" }),
		class_name = string.format(
			"icon %s",
			string.find(n.app_name, "-symbolic", 0, true) ~= nil and "symbolic" or ""
		),
		valign = "CENTER",
		halign = "START",
	})

	local image_path

	if n:get_hint("image-path") then
		image_path = n:get_hint("image-path"):get_data_as_bytes():get_data()
	else
		image_path = n.app_icon
	end

	local image = Widget.Icon({
		css = string.format("font-size: %.5frem", rem(70)),
		class_name = "image",
		valign = "CENTER",
		visible = image_path and #image_path > 1,
		icon = (image_path and #image_path > 1) and image_path or nil,
	})

	return Widget.EventBox({
		on_click_release = function(this)
			this:destroy()
		end,
		Widget.Box({
			orientation = "VERTICAL",
			Widget.Box({
				class_name = "notification-title",
				spacing = 10,
				app_icon,
				app_name,
				Widget.Icon({
					icon = "window-close-symbolic",
					class_name = "icon",
					halign = "END",
					hexpand = true, -- Ensure the icon can expand horizontally
				}),
			}),
			Widget.Box({
				spacing = 10,
				image,
				Widget.Box({
					orientation = "VERTICAL",
					class_name = "notification-box",
					spacing = 10,
					summary,
					body,
				}),
			}),
		}),
	})
end

-- Dynamically fetch Adwaita theme background color
local function get_theme_background()
	local widget = Gtk.Window()
	local context = widget:get_style_context()
	local color = context:get_background_color(Gtk.StateFlags.NORMAL)
	widget:destroy()
	return string.format("#%02x%02x%02x", color.red * 255, color.green * 255, color.blue * 255)
end

local notif_background_color = get_theme_background()

return function(gdkmonitor)
	return Widget.Window({
		class_name = "notifications",
		visible = false,
		anchor = Astal.WindowAnchor.TOP,
		width_request = math.floor(gdkmonitor:get_geometry().width / 3),
		setup = function(self)
			local count = 0
			self:hook(notifd, "notified", function()
				count = count + 1
				self.visible = true
			end)
			self:hook(notifd, "resolved", function()
				count = count - 1
				if count == 0 then
					timeout(popup_timeout, function()
						self.visible = false
					end)
				end
			end)
		end,
		Widget.Box({
			setup = function(self)
				self:hook(notifd, "notified", function(_, id)
					if #self.children > 1 then
						self.children[1]:destroy()
					end
					local n = notifd:get_notification(id)
					local e_timeout = n.expire_timeout > 0 and n.expire_timeout * 1000 or popup_timeout
					local widget = notif_item(n)
					self:add(widget)
					timeout(e_timeout, function()
						widget:destroy()
					end)
				end)
			end,
			css = string.format("background-color: %s;", notif_background_color)
				.. string.format("padding: %.5frem", rem(10)),
			orientation = "VERTICAL",
			class_name = "notifications-container",
			spacing = 10,
		}),
	})
end
